incubator/virtualcluster/pkg/syncer/controllers/event/uws.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package event
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+
+	"github.com/kubernetes-sigs/multi-tenancy/incubator/virtualcluster/pkg/syncer/conversion"
+)
+
+// StartUWS starts the upward syncer
+// and blocks until an empty struct is sent to the stop channel.
+func (c *controller) StartUWS(stopCh <-chan struct{}) error {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("starting pod upward syncer")
+
+	if !cache.WaitForCacheSync(stopCh, c.eventSynced, c.nsSynced) {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	klog.V(5).Infof("starting workers")
+	for i := 0; i < c.workers; i++ {
+		go wait.Until(c.run, 1*time.Second, stopCh)
+	}
+	<-stopCh
+	klog.V(1).Infof("shutting down")
+
+	return nil
+}
+
+// run runs a run thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (c *controller) run() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *controller) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	klog.Infof("back populate event %+v", key)
+	err := c.backPopulate(key.(string))
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("error processing event %v (will retry): %v", key, err))
+	c.queue.AddRateLimited(key)
+	return true
+}
+
+func (c *controller) backPopulate(key string) error {
+	pNamespace, pName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key %v: %v", key, err))
+		return nil
+	}
+
+	pEvent, err := c.eventLister.Events(pNamespace).Get(pName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("could not find pEvent %s/%s in controller cache: %v", pNamespace, pName, err)
+	}
+
```

Comment:
Should we check pEvent.nvolvedObject.Kind here to filter all non-Pod related events earlier?